File: user-service/src/main/java/OrangeCloud/UserRepo/entity/Attachment.java
package OrangeCloud.UserRepo.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Attachment 엔티티
 * 프로필 이미지 등의 파일 첨부를 관리합니다.
 * 임시 파일(TEMP)과 확정 파일(CONFIRMED) 상태를 구분하여 관리합니다.
 */
@Entity
@Table(name = "attachments", indexes = {
        @Index(name = "idx_attachments_entity", columnList = "entityType, entityId"),
        @Index(name = "idx_attachments_entity_id", columnList = "entityId"),
        @Index(name = "idx_attachments_status", columnList = "status"),
        @Index(name = "idx_attachments_uploaded_by", columnList = "uploadedBy"),
        @Index(name = "idx_attachments_expires_at", columnList = "expiresAt")
})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class Attachment {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id", columnDefinition = "UUID")
    private UUID id;

    @Enumerated(EnumType.STRING)
    @Column(name = "entityType", length = 50, nullable = false)
    private EntityType entityType;

    @Column(name = "entityId", columnDefinition = "UUID")
    private UUID entityId;  // Nullable - 임시 파일은 아직 entity에 연결되지 않음

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20, nullable = false)
    @Builder.Default
    private AttachmentStatus status = AttachmentStatus.TEMP;

    @Column(name = "fileName", length = 255, nullable = false)
    private String fileName;

    @Column(name = "fileUrl", columnDefinition = "TEXT", nullable = false)
    private String fileUrl;

    @Column(name = "fileSize", nullable = false)
    private Long fileSize;

    @Column(name = "contentType", length = 100, nullable = false)
    private String contentType;

    @Column(name = "uploadedBy", columnDefinition = "UUID", nullable = false)
    private UUID uploadedBy;

    @Column(name = "expiresAt")
    private LocalDateTime expiresAt;  // 임시 파일의 만료 시간

    @CreationTimestamp
    @Column(name = "createdAt", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updatedAt")
    private LocalDateTime updatedAt;

    @Column(name = "deletedAt")
    private LocalDateTime deletedAt;  // Soft delete

    /**
     * EntityType enum
     */
    public enum EntityType {
        USER_PROFILE
    }

    /**
     * AttachmentStatus enum
     */
    public enum AttachmentStatus {
        TEMP,       // 임시 상태
        CONFIRMED   // 확정 상태
    }

    /**
     * 임시 첨부파일 생성
     */
    public static Attachment createTemp(
            EntityType entityType,
            String fileName,
            String fileUrl,
            Long fileSize,
            String contentType,
            UUID uploadedBy,
            LocalDateTime expiresAt) {
        return Attachment.builder()
                .entityType(entityType)
                .entityId(null)  // 임시 파일은 entity에 연결되지 않음
                .status(AttachmentStatus.TEMP)
                .fileName(fileName)
                .fileUrl(fileUrl)
                .fileSize(fileSize)
                .contentType(contentType)
                .uploadedBy(uploadedBy)
                .expiresAt(expiresAt)
                .build();
    }

    /**
     * 첨부파일 확정 (TEMP -> CONFIRMED)
     */
    public void confirm(UUID entityId) {
        this.entityId = entityId;
        this.status = AttachmentStatus.CONFIRMED;
        this.expiresAt = null;  // 확정되면 만료 시간 제거
    }

    /**
     * Soft delete
     */
    public void softDelete() {
        this.deletedAt = LocalDateTime.now();
    }

    /**
     * 만료 여부 확인
     */
    public boolean isExpired() {
        return status == AttachmentStatus.TEMP 
                && expiresAt != null 
                && LocalDateTime.now().isAfter(expiresAt);
    }
}

File: board-service/internal/domain/attachment.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// EntityType represents the type of entity an attachment is associated with
type EntityType string

const (
	EntityTypeBoard   EntityType = "BOARD"
	EntityTypeProject EntityType = "PROJECT"
	EntityTypeComment EntityType = "COMMENT"
)

// AttachmentStatus represents the status of an attachment
type AttachmentStatus string

const (
	AttachmentStatusTemp      AttachmentStatus = "TEMP"      // Temporary status
	AttachmentStatusConfirmed AttachmentStatus = "CONFIRMED" // Confirmed status
)

// Attachment represents a file attachment associated with a board or project
// This is a polymorphic relationship - EntityID can reference Board, Project, or Comment
// ⚠️ IMPORTANT: Do not add foreign key constraints on EntityID as it references multiple tables
type Attachment struct {
	BaseModel
	EntityType  EntityType       `gorm:"type:varchar(50);not null;index:idx_attachments_entity,priority:1" json:"entity_type"`
	EntityID    *uuid.UUID       `gorm:"type:uuid;index:idx_attachments_entity,priority:2" json:"entity_id"` // ✅ FK 제거, 다형성 관계
	Status      AttachmentStatus `gorm:"type:varchar(20);not null;default:'TEMP';index:idx_attachments_status" json:"status"`
	FileName    string           `gorm:"type:varchar(255);not null" json:"file_name"`
	FileURL     string           `gorm:"type:text;not null" json:"file_url"` // ✅ S3 key만 저장 (full URL 아님)
	FileSize    int64            `gorm:"not null" json:"file_size"`
	ContentType string           `gorm:"type:varchar(100);not null" json:"content_type"`
	UploadedBy  uuid.UUID        `gorm:"type:uuid;not null;index:idx_attachments_uploaded_by" json:"uploaded_by"`
	ExpiresAt   *time.Time       `gorm:"type:timestamp;index:idx_attachments_expires_at" json:"expires_at"`
}

// TableName specifies the table name for Attachment
func (Attachment) TableName() string {
	return "attachments"
}
